File: testLib/src/main/java/me/ycdev/android/lib/test/log/TimberJvmTree.java
package me.ycdev.android.lib.test.log;

import java.util.ArrayList;

import androidx.annotation.NonNull;

import timber.log.Timber;

public class TimberJvmTree extends Timber.Tree {
    private ArrayList<String> mLogs = new ArrayList<>();

    public void clear() {
        mLogs.clear();
    }

    public boolean hasLogs() {
        return !mLogs.isEmpty();
    }

    @Override
    protected void log(int priority, String tag, @NonNull String message, Throwable t) {
        String log = AndroidLogHelper.getPriorityName(priority) + "/" + tag + ": " + message;
        mLogs.add(log);
        System.out.println(log);
        if (t != null) {
            t.printStackTrace(System.out);
        }
    }
}

File: testLib/src/main/java/me/ycdev/android/lib/test/log/TimberJvmTree.kt
package me.ycdev.android.lib.test.log

import java.util.ArrayList
import timber.log.Timber

@Suppress("unused")
class TimberJvmTree : Timber.Tree() {
    private var logs: ArrayList<String>? = null

    fun clear() {
        logs?.clear()
    }

    fun hasLogs(): Boolean {
        return logs?.isNotEmpty() ?: false
    }

    fun keepLogs() {
        logs = ArrayList()
    }

    override fun log(priority: Int, tag: String?, message: String, t: Throwable?) {
        val log = AndroidLogHelper.getPriorityName(priority) + "/" + tag + ": " + message
        logs?.add(log)
        println(log)
        t?.printStackTrace(System.out)
    }

    companion object {
        fun plantIfNeeded() {
            // only plant TimberJvmTree once
            Timber.forest().forEach {
                if (it is TimberJvmTree) {
                    return
                }
            }
            Timber.plant(TimberJvmTree())
        }
    }
}
